clients/src/main/java/org/apache/kafka/clients/producer/KafkaProducer.java

```diff
@@ -479,6 +479,12 @@ private static int parseAcks(String acksString) {
             // handling exceptions and record the errors;
             // for API exceptions return them in the future,
             // for other exceptions throw directly
+        } catch (BufferExhaustedException e) {
+            this.errors.record();
+            this.metrics.sensor("buffer-exhausted-records").record();
+            if (this.interceptors != null)
+                this.interceptors.onSendError(record, tp, e);
+            throw e;
```

Comment:
This is a good point. The current implementation is a bit misleading with regards to the javadoc which states:

```java
* @throws TimeoutException If the time taken for fetching metadata or allocating memory for the record has surpassed <code>max.block.ms</code>.
```

We don't actually throw that exception unless you do `Future.get`. It seems to me that the change in this PR actually fixes the implementation to match the specified contract.

Another option is to change the javadoc to match the implementation (probably less likely to break users) and then we would simply have a check in the `ApiException` catch block to record the metric for `BufferExhaustedException`. This seems safer.

Thoughts?